Windows 11 22H2: add good UIA window class names for system tray overflow window and Open With dialog to allow mouse and touch interaction

josephsl opened the pull request:
### Link to issue number:
Closes #14538 
Closes #14539 

### Summary of the issue:
Mouse and/or touch interaction does not work properly for Windows 11 22H2 UI, including upcoming redesigned system tray overflow window and Open With dialog.

### Description of user facing changes
Mouse and touch interaction will work when opening system tray overflow window and Open With dialog in Windows 11 2022 Update.

### Description of development approach
Edited "isGoodUIAWindow" found in File Explorer and Open With app modules to classify the following window class names as good UIA windows:

* File Explorer: TopLevelWindowForOverflowXamlIsland (upcoming redesigned system tray overflow window)
* Open With: Open With (top-level window for the dialog)

Along with this change, type information were added to these methods.

### Testing strategy:
Manual testing (prereq: mouse and/or touchscreen; for mouse, mouse tracking should be turned on and NVDA sould be told to follow mouse from review cursor settings panel):

Redesigned system tray overflow window (Windows 11 22H2 build 22623 and later): press Windows +B to move focus to "show hidden icons" button, press Space to expand it, then use the mouse to move around the overflow window (located above the icons button).

Open With dialog: click on a file type that Windows cannot open (such as a pickle file) and use the mosue and/or touch to navigate and interact with the dialog.

### Known issues with pull request:
None

### Change log entries:
Bug fixes (under Windows 11 section):

In Windows 11 22H2, it is again possible to use mouse and touch interaction to navigate and interact in areas such as redesigned system tray overflow window and Open With dialog. (#14538, #14539)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

josephsl committed:
appModules/openwith: recognize 'Open With' window class as a good UIA window. Re #14538.

Windows 11 22H2 (at least newer updates) include modernized Open With dialog powered by WinUI, meaning no proper mouse and/or touch interaction. Therefore treat the top window for this dialog as a good UIA window so mouse and touch interaction can work properly. Also, rewrite isGoodUIAWindow method with one window class name per line.

josephsl committed:
appModules/openwith: add typing information for isGoodUIAWindow method

josephsl committed:
appModules/explorer: add 'TopLevelWindowForOverflowXamlIsland' as a good UIA window. Re #14539.

Later revisions of Windows 11 Version 22H2 (2022 Update) comes with redesigned system tray overflow window (class name: TopLevelWindowForOverflowXamlIsland) but without proper mouse and touch interaction from NVDA. Therefore, add the new class as a good UIA window so mouse and touch interaction can work.

josephsl committed:
appModules/explorer: add typing information for isGoodUIAWindow method.

josephsl committed:
appModules/explorer and openwith: update copyright year

mazen428 commented:
Hello.
All the windows that block touch or mouse access seem to have the name 'DesktopWindowXamlSource', with a role of unknown or pane. I think it would be better to make a window like this UIA or just transparent to mouse instead of manually adding class names for everything. What about apps from the Microsoft Store that use XAML.

josephsl commented:
Hi, tried this to no avail at the moment – it does not work for touch users, although further testing is needed to determine what’s going on. Thanks.

josephsl commented:
Hi,

Just telling NVDA that XAML source windows are UIA won't be enough - it is part of aps (specifically, WinUI/XAML Islands), and it is really the top-level windows that are seen as IAccessible, thus these windows must be the ones reclassified as UIA. We can't also rely on the idea that XAML source windows will have the same window class names across apps - class names are different between File Explorer and Windows Terminal. Most importantly, if XAML source window is reclassified as a UIA element, it leads to NVDA becoming unresponsive at startup for several seconds.

Thanks.

seanbudd commented on source/appModules/explorer.py:
```diff
@@ -504,7 +504,7 @@ def event_gainFocus(self, obj, nextHandler):
 		nextHandler()
 
-	def isGoodUIAWindow(self, hwnd):
+	def isGoodUIAWindow(self, hwnd: int) -> bool:
```
```suggestion
	from winAPI.types import HWNDValT
	def isGoodUIAWindow(self, hwnd: HWNDValT) -> bool:
```

seanbudd commented on source/appModules/openwith.py:
```diff
@@ -40,9 +40,14 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
-		# #11335: Open With dialog isn't read in Windows 10 Version 2004 (May 2020 Update).
-		# Note that treating the below window as a UIA window will make NVDA no longer announce "pane".
-		if winUser.getClassName(hwnd) == "Shell_Flyout":
+	def isGoodUIAWindow(self, hwnd: int) -> bool:
```
same here in regards to the hwnd type

josephsl commented:
Hi, ah, wasn’t sure which type to apply for window handle… Thanks.

josephsl committed:
appModules/explorer: hwnd: int -> hwnd type

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

josephsl committed:
appModules/openwith: hwnd: int -> hwnd type

seanbudd commented on source/appModules/openwith.py:
```diff
@@ -40,9 +40,15 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
-		# #11335: Open With dialog isn't read in Windows 10 Version 2004 (May 2020 Update).
-		# Note that treating the below window as a UIA window will make NVDA no longer announce "pane".
-		if winUser.getClassName(hwnd) == "Shell_Flyout":
+	from winAPI.types import HWNDValT
```
You should probably pull these imports out to the top of the file

josephsl committed:
appModules/explorer and openwith: move winAPI import to top of the module.

Suggested by Sean Budd (NV Access): move winAPI import to the top of the modules.

seanbudd committed:
Merge remote-tracking branch 'origin/master' into win11-22h2ExplorerOpenWithGoodUIAWindow

seanbudd committed:
update changes

Rockstar50373 commented:
Thanks for the update!

On Sun, Jan 15, 2023, 6:58 PM Sean Budd ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In source/appModules/openwith.py
> <https://github.com/nvaccess/nvda/pull/14540#discussion_r1070749844>:
>
> > @@ -40,9 +40,15 @@ def chooseNVDAObjectOverlayClasses(self, obj, clsList):
>  			elif automationId == "ImmersiveOpenWithFlyout":
>  				clsList.insert(0, ImmersiveOpenWithFlyout)
>
> -	def isGoodUIAWindow(self, hwnd):
> -		# #11335: Open With dialog isn't read in Windows 10 Version 2004 (May 2020 Update).
> -		# Note that treating the below window as a UIA window will make NVDA no longer announce "pane".
> -		if winUser.getClassName(hwnd) == "Shell_Flyout":
> +	from winAPI.types import HWNDValT
>
> You should probably pull these imports out to the top of the file
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/14540#pullrequestreview-1249433408>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXT2QUIVEZQAU7VO7FWZUCDWSSTMPANCNFSM6AAAAAAT3PQB5Q>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
